Support symlinks, proper time and translate errors

Crementif opened the pull request:
It has some duplicate wut structs to extend them, so I'll probably make a pull request to add it to wut (might take a while though). But let me know if there's any issues or things I should look into.

I tried to make everything backwards compatible as much as possible while being sensible. Existing code should still compile (from what I've tested).

I had some issues deciding whether I wanted to return multiple file modes, but it's not really clear whether it's really supported in posix and due to there also being a conversion to DT_REG/DT_DIR etc for directory entries it's probably just better to have no mixing. Also better for compatibility reasons since existing code probably doesn't use bit operators to check them.

Credits to @koolkdev for figuring out the FSStat flags and structs.

Crementif committed:
Support time, fix modes and translate errors

Thanks to @koolkdev for figuring out the FSStat flags and structs.
Thanks to WUT for the error codes.

Maschell commented on source/iosuhax_devoptab.c:
```diff
@@ -104,14 +110,71 @@ static char *fs_dev_real_path(const char *path, fs_dev_private_t *dev) {
     return new_name;
 }
 
+static int fs_dev_translate_error(FSStatus error) {
```
Mixing FS_STATUS and FS_ERROR doesn't really make much sense, see https://github.com/devkitPro/wut/pull/213

Crementif replied:
Added the changes from that PR!

Maschell commented on source/iosuhax_devoptab.c:
```diff
@@ -762,17 +880,32 @@ static int fs_dev_dirnext_r(struct _reent *r, DIR_ITER *dirState, char *filename
         memset(st, 0, sizeof(struct stat));
-        st->st_mode   = (dir_entry->stat.flag & 0x80000000) ? S_IFDIR : S_IFREG;
+
+        // Turn Wii U's FSStat mode to posix stat
```
This mapping is done three times in the code, it should/could extracted into a own function like for example: `fs_dev_translate_error`

Crementif replied:
Yup, definitely makes sense. Done!

Maschell commented on include/iosuhax.h:
```diff
@@ -29,27 +29,53 @@
+#define DIR_ENTRY_IS_DIRECTORY   FS_DEV_STAT_DIRECTORY
+#endif
+
+typedef enum FSDevMode {
```
Seems to be identical to [FSMode](https://github.com/devkitPro/wut/blob/master/include/coreinit/filesystem.h#L121) from wut. I would prefer using the existing modes instead of defining a new one (or just do a typedef?)

Crementif replied:
Didn't want to include a pretty sizeable header like coreinit/filesystem.h for just the fsmode, but it's probably something that'll be done in the future when the new flags are added to wut so it doesn't really hurt.

Removed the local definition of it.

Maschell commented on source/iosuhax_devoptab.c:
```diff
@@ -28,11 +28,16 @@
 
+#include <coreinit/time.h>
+#include <coreinit/filesystem.h>
+void FSTimeToCalendarTime(OSTime time, OSCalendarTime *calendarTime);
```
I would prefer adding this to wut instead of defining it just in this project

Crementif replied:
Yeah, I plan on doing some pull requests with the changes I've made here regarding the time function and the dev flags. You probably know better if waiting with this PR if waiting for that is a better idea. When writing I was just thinking that it'd probably take a while for that stuff to be in a stable wut release, so I opted for defining it in the .cpp file.

Maschell commented on source/iosuhax_devoptab.c:
```diff
@@ -104,14 +110,71 @@ static char *fs_dev_real_path(const char *path, fs_dev_private_t *dev) {
+static time_t fs_dev_translate_time(OSTime timeValue) {
+    OSCalendarTime fileTime;
+    FSTimeToCalendarTime(timeValue, &fileTime);
+    // WHBLogPrintf("[DEBUG] date: %d-%d-%d, time: %d:%d:%d", fileTime.tm_year, fileTime.tm_mon, fileTime.tm_mday, fileTime.tm_hour, fileTime.tm_min, fileTime.tm_sec);
```
Remove comment

Maschell commented on source/iosuhax_devoptab.c:
```diff
@@ -591,13 +679,43 @@ static int fs_dev_chmod_r(struct _reent *r, const char *path, int mode) {
     return 0;
 }
 
+// static int fs_dev_fchmod_r(struct _reent *r, void *fd, mode_t mode) {
```
Why is this commented out? Please remove it or implement/test it properly. (Maybe move it into a new WIP issue if it's not ready yet?)

Crementif replied:
It does work, but it required tracking the path inside the file handle dir and I think there must be a better solution. I eventually wanna fix some of the unsupported functions in this and wut's SD devoptab, but decided that's better to leave out in this PR.

I've removed all the code related to it for now though.

Maschell commented:
Thanks for the contribution.
1. Please fix the the formatting according to the ,clangformat, can be easily done via docker (`docker run --rm -v ${PWD}:/src wiiuenv/clang-format:13.0.0-2 -r ./source ./include -i`)
2. I've only done a code review, but didn't test it yet. Beside the parts I have commented it looks reasonable though, great stuff!

Crementif commented:
I've made the changes and also ran the clangformat. Thanks for the quick review.

Let me know if you need any help with the testing. I basically just modified WiiUFTPServer to read out the information, still got a [video](https://streamable.com/q8p2ii) of some of the results while going to the BotW update which has the Movie and Layout movies that are symlinks. [Here's the modified version](https://github.com/wiiu-env/libiosuhax/files/8384977/WiiUFtpServer.zip) if you wanna check something yourself, might save some time that shows the file flags.

I also send it to Laf111 and he tested it more properly and said that the timestamps only appear to be correct for folders that are deeper into the mlc filesystem (usually, the levels after /title, /save). Maybe the directories that also have the quota flag? No clue honestly. Does work on all of the folders on the usb01 filesystem.  It also returns 1980 when it doesn't have a timestamp apparently.

He also mentioned that he still can't faccess a file or use realpath/readlink, which makes sense with the changes I've done. It only gives the correct symlink type back.

I also added a bit more of an explanation to the new flags in the changes.

Crementif commented:
I opened up a PR at WUT to upstream some of the local changes done: https://github.com/devkitPro/wut/pull/215.

Crementif commented:
I've removed the code that is now upstreamed to wut's source, so this should be good to merge. Will need a new devkitpro release or to build it from source, so I've also added a disclaimer to the readme.

Maschell commented:
Thanks! I updated the Dockerfiles to use the latest wut and will test it (hopefully) later today

Crementif commented:
I updated it to now also reflect the permissions, and also fixed the order bug that Gary spotted where chmod was uses as devicedata. Apparently it's always been there? 

Also added lstat support to the devoptab, but the (f)stat both require fixing since they've always been broken.

Maschell commented on source/iosuhax_devoptab.c:
```diff
@@ -762,17 +877,21 @@ static int fs_dev_dirnext_r(struct _reent *r, DIR_ITER *dirState, char *filename
+        st->st_rdev    = st->st_dev;
+        st->st_size    = dir_entry->info.size;
+        st->st_blksize = 512;
+        st->st_blocks  = (st->st_size + st->st_blksize - 1) >> 9;
```
Where is that 9 coming? This looks like a hardcoded blocksize of 512? 
Is the blocksize **always** 512?

doing `st->st_blocks  = (st->st_size + st->st_blksize - 1) / st->st_blksize;` feels a bit better (even though the st_blksize is hardcoded anyway).

Maschell commented on source/iosuhax_devoptab.c:
```diff
@@ -104,6 +105,99 @@ static char *fs_dev_real_path(const char *path, fs_dev_private_t *dev) {
+    return (int) error;
+}
+
+static mode_t fs_dev_translate_mode(FSStat fileStat, bool followLinks, bool isRootDirectory) {
```
Can you add curly braces to all if/else conditions in this function?

Maschell commented on source/iosuhax_devoptab.c:
```diff
@@ -335,56 +429,80 @@ static int fs_dev_fstat_r(struct _reent *r, void *fd, struct stat *st) {
+    st->st_rdev    = st->st_dev;
+    st->st_size    = stats.size;
+    st->st_blksize = 512;
+    st->st_blocks  = (st->st_size + st->st_blksize - 1) >> 9;
```
See the other comment

Maschell commented on source/iosuhax_devoptab.c:
```diff
@@ -335,56 +429,80 @@ static int fs_dev_fstat_r(struct _reent *r, void *fd, struct stat *st) {
+    st->st_rdev    = st->st_dev;
+    st->st_size    = stats.size;
+    st->st_blksize = 512;
+    st->st_blocks  = (st->st_size + st->st_blksize - 1) >> 9;
```
see other comment

Maschell commented on source/iosuhax_devoptab.c:
```diff
@@ -403,34 +521,31 @@ static int fs_dev_stat_r(struct _reent *r, const char *path, struct stat *st) {
+    st->st_rdev    = st->st_dev;
+    st->st_size    = stats.size;
+    st->st_blksize = 512;
+    st->st_blocks  = (st->st_size + st->st_blksize - 1) >> 9;
```
^

Maschell reviewed (changes requested):
Just a few more things inside the code

Crementif commented:
Changed everything regarding the feedback. I think the 9 is indeed just to divide by 512 for optimization reasons, but it probably gets optimized anyway. Also accidentally pushed some commits that I made to bisect some stuff, so hence the force push.

Crementif commented:
I added the changes mentioned by GaryOderNights/you in the discord, and I also updated the conversion from CafeOS bits to Unix bits ([tested them here](https://www.onlinegdb.com/TUREK7exS) to make sure I didn't screw up any bitshifts).

Is it truly time to get this merged?!?!

Maschell committed:
Force set the FS_STAT_FILE flag if the filesize is > 0
